### PR TITLE
Prodigy optimizer: parameter-free LR adaptation for DrivAerML

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis: Prodigy Optimizer — parameter-free LR adaptation to escape the exhausted LR search

**Tag:** \`radford-20260421-relaunch\`

### Motivation

The DrivAerML LR sweep (1e-4 to 1e-3) has been run extensively and shows a frustratingly flat landscape — 5e-4 is the best but the gap to 3.71% persists. Prodigy (Mishchenko & Defazio, ICML 2024) is a parameter-free optimizer that learns its own effective step size online using D-Adaptation. It maintains per-parameter effective learning rates that adapt to the local loss landscape curvature. On DrivAerML, where the 3D surface topology creates an extremely ill-conditioned optimization landscape, Prodigy's adaptive scaling could find descent directions that no fixed LR schedule reaches. Prodigy has been validated across ViTs, LLMs, and diffusion models — it consistently matches or beats hand-tuned Adam within the same epoch budget.

### Code Changes Required

```bash
pip install prodigyopt
```

Add \`--optimizer prodigy\` in the optimizer factory in \`train.py\`:

\`\`\`python
from prodigyopt import Prodigy

if args.optimizer == "prodigy":
    optimizer = Prodigy(
        model.parameters(),
        lr=1.0,           # Prodigy convention — it ignores this
        weight_decay=args.weight_decay,
        safeguard_warmup=True,   # CRITICAL: prevents early divergence
        use_bias_correction=True,
    )
\`\`\`

**Important:** Prodigy manages its own effective LR internally. The \`lr=1.0\` is a convention. The \`--lr\` flag should be ignored when using Prodigy. Keep the cosine scheduler — Prodigy composes with external schedulers (they modulate Prodigy's adapted base LR).

### Current Baselines (MUST BEAT)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| DrivAerML | \`val_primary/surface_rel_l2_pct\` | **4.619%** | #2691 |
| TandemFoil | \`val_primary/surface_pressure_mae\` | **26.134** | #2899 |
| TandemFoil Paper | \`val_primary/field_mse\` | not yet established | — |
| AirfRANS | \`val_primary/surface_mse\` | **0.000727** | #2899 |

### Run Matrix (8 GPUs)

Use \`--wandb_group radford-20260421-relaunch-prodigy\` for all runs.

See the supplementary comment below for the full 8-GPU run matrix with exact commands.

### What to Report

For each run: primary val metric at best checkpoint, test metric at best val checkpoint, Prodigy's effective LR over training (log it via \`optimizer.param_groups[0]['d']\` — this is Prodigy's adapted step size). Note convergence speed vs AdamW/Lion baselines.

### Key Question

Can Prodigy's adaptive LR find a better optimization trajectory than our hand-tuned lr=5e-4 for DrivAerML? The "no cosine" variant tests whether Prodigy's internal scheduling is sufficient alone.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Supplementary: Full 8-GPU Run Matrix

Use \`--wandb_group radford-20260421-relaunch-prodigy\` for all runs.

**GPU 0 — DrivAerML: Prodigy at 4L/512d baseline**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer prodigy --lr 1.0 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 1 — DrivAerML: Prodigy + gc=1.0**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer prodigy --lr 1.0 --grad-clip 1.0 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 2 — DrivAerML: Prodigy + WD=1e-2 + gc=1.0**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer prodigy --lr 1.0 --weight-decay 1e-2 --grad-clip 1.0 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 3 — DrivAerML: Prodigy no cosine (pure Prodigy scheduling)**
\`\`\`bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer prodigy --lr 1.0 --cosine-t-max 99999 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
\`\`\`

**GPU 4 — TandemFoil: Prodigy at 3L/192d**
\`\`\`bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer prodigy --lr 1.0 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 5 — TandemFoil Paper: Prodigy**
\`\`\`bash
cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer prodigy --lr 1.0 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 6 — AirfRANS: Prodigy at 2L/256d**
\`\`\`bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer prodigy --lr 1.0 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999
\`\`\`

**GPU 7 — AirfRANS: Prodigy no cosine (pure Prodigy)**
\`\`\`bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer prodigy --lr 1.0 --cosine-t-max 99999 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999
\`\`\`
